### PR TITLE
chore: pedersen hash in Noir

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Related to  #4931 

## Summary\*
Noir implementation for Pedersen Hash.

## Additional Context
I added a derive_generators() built-in function, which only works for bn254, and then use pedersen commitments and scalar mul.

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: noir_stdlib/src/hash.nr
Comment (guipublic) [reply noir_stdlib/src/hash.nr:53]: Do you mean `"pedersen_hash_length"` as a comptime global?

File: compiler/noirc_evaluator/src/ssa/ir/instruction/call.rs
Comment (guipublic) [reply compiler/noirc_evaluator/src/ssa/ir/instruction/call.rs]: Done

File: compiler/noirc_evaluator/src/ssa/ir/instruction/call.rs
Comment (guipublic) [reply compiler/noirc_evaluator/src/ssa/ir/instruction/call.rs]: Done

File: noir_stdlib/src/hash.nr
Comment (guipublic) [reply noir_stdlib/src/hash.nr]: Done

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs
Comment (guipublic) [reply compiler/noirc_evaluator/src/ssa/ir/instruction.rs]: Done

File: compiler/noirc_evaluator/src/ssa/ir/instruction/call.rs
Comment (guipublic) [reply compiler/noirc_evaluator/src/ssa/ir/instruction/call.rs]: Done

File: acvm-repo/bn254_blackbox_solver/src/generator/mod.rs
Comment (guipublic) [reply acvm-repo/bn254_blackbox_solver/src/generator/mod.rs:7]: Done

File: acvm-repo/bn254_blackbox_solver/src/lib.rs
Comment (guipublic) [reply acvm-repo/bn254_blackbox_solver/src/lib.rs]: Done

File: acvm-repo/bn254_blackbox_solver/src/lib.rs
Comment (TomAFrench) [inline acvm-repo/bn254_blackbox_solver/src/lib.rs:61]:
```suggestion
        let result = if let (x, y) = result.xy() {
            (FieldElement::from_repr(*x), FieldElement::from_repr(*y))
        } else {
            (FieldElement::zero(), FieldElement::zero())
        };
        
        Ok(result)
```

File: noir_stdlib/src/hash.nr
Comment (TomAFrench) [inline noir_stdlib/src/hash.nr:64]:
```suggestion
#[field(bn254)]
fn derive_generators<N, M>(
    domain_separator_bytes: [u8; M],
    starting_index: u32
) -> [EmbeddedCurvePoint; N] {
    crate::assert_constant(domain_separator_bytes);
    crate::assert_constant(starting_index);
    __derive_generators(domain_separator_bytes, N, starting_index)
}
```

`num_generators` is redundant here.

File: noir_stdlib/src/hash.nr
Comment (TomAFrench) [inline noir_stdlib/src/hash.nr:72]: Not sure how easy it would be to pull values of the generics from inside of the compiler (so we can remove `num_generators` from here too)

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/ir/instruction.rs:96]: The enum variant name hasn't been updated here.

**Review by TomAFrench — APPROVED** [review]
some nits and then ready to go

File: acvm-repo/bn254_blackbox_solver/src/lib.rs
Comment (guipublic) [reply acvm-repo/bn254_blackbox_solver/src/lib.rs]: Done (but not using the zero trait which is not in scope).

File: noir_stdlib/src/hash.nr
Comment (guipublic) [reply noir_stdlib/src/hash.nr]: Done

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs
Comment (guipublic) [reply compiler/noirc_evaluator/src/ssa/ir/instruction.rs]: Done

File: noir_stdlib/src/hash.nr
Comment (guipublic) [reply noir_stdlib/src/hash.nr]: Done, I assumed the result is a fixed array.

**Comment by TomAFrench** [comment]
Checking with ilyas that this isn't going to break the AVM before we merge.

**Review by TomAFrench — CHANGES_REQUESTED** [review]
From @IlyasRidhuan:
> right! so it will break everything because loads of contracts have storage mappings that use pedersen hashing.
> Can you hold off on syncing until friday (best day to try a sync of course) and ill get the blackboxes added in the simulator/transpiler/prover on our side

We should probably just hold off from merging this on our side while AVM builds out support for "pedersen as MSM"

**Comment by guipublic** [comment]
> From @IlyasRidhuan:
> 
> > right! so it will break everything because loads of contracts have storage mappings that use pedersen hashing.
> > Can you hold off on syncing until friday (best day to try a sync of course) and ill get the blackboxes added in the simulator/transpiler/prover on our side
> 
> We should probably just hold off from merging this on our side while AVM builds out support for "pedersen as MSM"

I am not sure why it would break anything, with this PR se simply do not use Pedersen hash opcode anymore, but we do not add/modify opcodes?

**Comment by TomAFrench** [comment]
The issue is that the AVM transpiler currently has handling for pedersen hash opcodes but not for MSM opcodes. If we replace all the hashes with commitment + MSM then the AVM won't know what to do with them. There's also a bunch of usage of pedersen in aztec contracts so this will break.

**Comment by IlyasRidhuan** [comment]
Paging in @dbanks12  or @fcarreiro, who are more across the nuances within the contracts

File: noir_stdlib/src/hash.nr
Comment (fcarreiro) [inline noir_stdlib/src/hash.nr:61]: Do we also need to add support for this on the AVM side? Not sure how builtins work. @sirasistant

Comment (TomAFrench) [reply noir_stdlib/src/hash.nr:61]: Nah, builtins are all applied by the compiler so this call will be replaced with an array literal containing the pedersen generators.

Comment (fcarreiro) [reply noir_stdlib/src/hash.nr:61]: Awesome thanks

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**
